controllers/awsmachine_controller.go

```diff
@@ -343,39 +366,75 @@ func (r *AWSMachineReconciler) reconcileNormal(ctx context.Context, machineScope
+		_, err = r.ensureTags(ec2svc, machineScope.AWSMachine, machineScope.GetInstanceID(), machineScope.AdditionalTags())
+		if err != nil {
+			return reconcile.Result{}, errors.Errorf("failed to ensure tags: %+v", err)
+		}
 	}
 
-	// TODO(vincepri): Remove this annotation when clusterctl is no longer relevant.
-	machineScope.SetAnnotation("cluster-api-provider-aws", "true")
+	// tasks that can only take place during operational instance states
+	if infrav1.InstanceOperationalStates.Has(string(instance.State)) {
+		machineScope.SetAddresses(instance.Addresses)
 
-	existingSecurityGroups, err := ec2svc.GetInstanceSecurityGroups(*machineScope.GetInstanceID())
-	if err != nil {
-		return reconcile.Result{}, err
-	}
+		if err := r.reconcileLBAttachment(machineScope, clusterScope, instance); err != nil {
+			return reconcile.Result{}, errors.Errorf("failed to reconcile LB attachment: %+v", err)
+		}
 
-	// Ensure that the security groups are correct.
-	_, err = r.ensureSecurityGroups(ec2svc, machineScope, machineScope.AWSMachine.Spec.AdditionalSecurityGroups, existingSecurityGroups)
-	if err != nil {
-		return reconcile.Result{}, errors.Errorf("failed to apply security groups: %+v", err)
-	}
+		existingSecurityGroups, err := ec2svc.GetInstanceSecurityGroups(*machineScope.GetInstanceID())
+		if err != nil {
+			return reconcile.Result{}, err
+		}
 
-	// Ensure that the tags are correct.
-	_, err = r.ensureTags(ec2svc, machineScope.AWSMachine, machineScope.GetInstanceID(), machineScope.AdditionalTags())
-	if err != nil {
-		return reconcile.Result{}, errors.Errorf("failed to ensure tags: %+v", err)
+		// Ensure that the security groups are correct.
+		_, err = r.ensureSecurityGroups(ec2svc, machineScope, machineScope.AWSMachine.Spec.AdditionalSecurityGroups, existingSecurityGroups)
+		if err != nil {
+			return reconcile.Result{}, errors.Errorf("failed to apply security groups: %+v", err)
+		}
 	}
 
 	return reconcile.Result{}, nil
 }
 
-func (r *AWSMachineReconciler) getOrCreate(scope *scope.MachineScope, ec2svc services.EC2MachineInterface) (*infrav1.Instance, error) {
+func (r *AWSMachineReconciler) deleteEncryptedBootstrapDataSecret(machineScope *scope.MachineScope, secretSvc services.SecretsManagerInterface) error {
+	// do nothing if there isn't asecret
+	if machineScope.GetSecretARN() == "" {
+		return nil
+	}
+	// the machine must not have failed
+	machineOK := machineScope.AWSMachine.Status.FailureReason == nil && machineScope.AWSMachine.Status.FailureMessage == nil
+
+	state := machineScope.GetInstanceState()
+	instanceOperational := state != nil && infrav1.InstanceOperationalStates.Has(string(*state))
+	notDeleted := machineScope.AWSMachine.ObjectMeta.DeletionTimestamp.IsZero()
+
+	// Do nothing if the AWSMachine is not in a failed state, and is operational from an EC2 perspective, but does not have a node reference
+	if machineOK && instanceOperational && machineScope.Machine.Status.NodeRef == nil && notDeleted {
+		return nil
+	}
```

Comment:
Can we rewrite this in a method or under Scope? I'm not totally sure about assigning boolean variables to expressions rather than having a check after another. In terms of readability, it's harder to read this way

Reply:
I don't see any other need for this logic, so I'm not sure putting it on the Scope makes sense?

Reply:
Could be useful to get `IsInstanceOperational` and `IsDeleted` as method on scope for other functions to use as well

Reply:
+1 to IsInstanceOperational. IsDeleted is trickier - are we talking about the Machine or the AWSMachine or either/both?